src/Lumper.Lib/BSP/Lumps/BspLumps/PakFileLump.cs

```diff
@@ -9,132 +10,201 @@ namespace Lumper.Lib.BSP.Lumps.BspLumps;
 using IO;
 using Lumps;
 using Newtonsoft.Json;
+using NLog;
 using SharpCompress.Archives.Zip;
 using SharpCompress.Common;
 using SharpCompress.Writers;
 using SharpCompress.Writers.Zip;
 using Struct;
 
-//[JsonConverter(typeof(PakFileJsonConverter))]
-public class PakFileLump : ManagedLump<BspLumpType>
+/// <summary>
+/// The pakfile lump (aka paklump) is just a zip archive for storing assets in the BSP.
+/// The overall archive is always uncompressed, but each item may be LZMA compressed.
+/// </summary>
+public class PakfileLump(BspFile parent) : ManagedLump<BspLumpType>(parent), IFileBackedLump
 {
     public List<PakfileEntry> Entries { get; private set; } = [];
 
-    private ZipArchive _zip;
     [JsonIgnore]
-    public ZipArchive Zip
-    {
-        get => _zip;
-        set
-        {
-            _zip = value;
-            UpdateEntries();
-        }
-    }
+    public Stream DataStream { get; set; } = null!;
+
+    public long DataStreamOffset { get; set; }
+
+    public long DataStreamLength { get; set; }
+
+    [JsonIgnore]
+    public ZipArchive Zip { get; private set; } = null!;
 
-    public override void Read(BinaryReader reader, long length)
+    private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
+
+    // Empty methods to avoid OO spaghetti, must call our Read/Write methods instead.
+    public override void Read(BinaryReader reader, long length) => throw new NotImplementedException();
+
+    public override void Write(Stream stream) => throw new NotImplementedException();
+
+    public void Read(BinaryReader reader, long length, IoHandler handler)
```

Comment:
I mainly did this with `Lump.Write` in mind, for which the gamelump and paklump versions must be passed a `DesiredCompression` since they use it internally. For other lumps, the compression is handler by `LumpWriter` (which calls `Lump.Write`). I give `Write` the signature `Write(Stream stream, DesiredCompression? compression, IoHandler? handler)`, but then I have to pass in `null` to everything besides the paklump and gamelump.

In general, I'd rather a base class not have a method signature where one of the params is only ever used by a minority of the sub classes, just feels ugly.

I suppose
```cs
public void Write(Stream stream) => Write(stream, null, null);
   
public abstract void Write(Stream stream, DesiredCompression? compression, IoHandler? handler);
```
isn't too bad?

Reply:
Nvm just made it use `IoHandler? handler = null` everywhere and just stomaching the endless null checks. I really don't like making IoHandler required, there's cases like EntityLumpViewModel which uses EntityLump's Write method to write the entity lump out to a stream, there's no need for an IoHandler/cancellation tokens in that case.